nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/ListenSMTP.java

```diff
@@ -0,0 +1,385 @@
+
+            @Override
+            public SSLSocket createSSLSocket(Socket socket) throws IOException {
+                InetSocketAddress remoteAddress = (InetSocketAddress) socket.getRemoteSocketAddress();
+
+                SSLSocketFactory socketFactory = finalSslContext.getSocketFactory();
+
+                SSLSocket s = (SSLSocket) (socketFactory.createSocket(socket, remoteAddress.getHostName(), socket.getPort(), true));
+
+                s.setUseClientMode(false);
+
+
+                // For some reason the createSSLContext above is not enough to enforce
+                // client side auth
+                // If client auth is required...
+                if (SSLContextService.ClientAuth.REQUIRED.toString().equals(context.getProperty(CLIENT_AUTH).getValue())) {
+                    s.setNeedClientAuth(true);
+                }
+
+
+                return s;
+            }
+        };
+
+
+        // Set some parameters to our server
+        server.setSoftwareName("Apache NiFi");
+
+
+        // Set the Server options based on properties
+        server.setPort(context.getProperty(SMTP_PORT).asInteger());
+        server.setHostName(context.getProperty(SMTP_HOSTNAME).getValue());
+        server.setMaxMessageSize(context.getProperty(SMTP_MAXIMUM_MSG_SIZE).asDataSize(DataUnit.B).intValue());
+        server.setMaxConnections(context.getProperty(SMTP_MAXIMUM_CONNECTIONS).asInteger());
+        server.setConnectionTimeout(context.getProperty(SMTP_TIMEOUT).asTimePeriod(TimeUnit.MILLISECONDS).intValue());
+
+
+        // Check if TLS should be enabled
+        if (sslContextService != null) {
+            server.setEnableTLS(true);
+        } else {
+            server.setHideTLS(true);
+        }
+
+        // Set TLS to required in case CLIENT_AUTH = required
+        if (SSLContextService.ClientAuth.REQUIRED.toString().equals(context.getProperty(CLIENT_AUTH).getValue())) {
+            server.setRequireTLS(true);
+        }
+
+        this.server = server;
+        server.start();
+
+        getLogger().info("Server started and listening on port " + server.getPort());
+
+        initialized.set(true);
+    }
+
+    @OnStopped
+    public void shutdown() throws Exception {
+        if (server != null) {
+            getLogger().debug("Shutting down server");
+            server.stop();
+            getLogger().info("Shut down {}", new Object[]{server});
+        }
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        try {
+            initializeSMTPServer(context);
+        } catch (Exception e) {
+            context.yield();
+            throw new ProcessException("Failed to initialize the SMTP server", e);
+        }
+
+        if (messages.isEmpty()) {
+            return;
+        }
+
+        FlowFile flowfile = session.create();
+        SmtpEvent message;
+
+        try {
+            message = messages.take();
```

Comment:
Oh, also this shouldn't "take". By 'take'ing it will wait until there is a message to take. I know there is the check to see if "messages" is empty in the event race condition occurs and the last message is removed, this will cause an unwanted hanging thread. "poll" would be better and then just return if it is null.